### ISSUE TITLE
Domains catch unrelated errors and mask real error sources

### ISSUE THREAD

**Issue by diversario** [BryanDonovan/node-cache-manager#28]
I just spent more than an hour trying to figure out why cache-manager was throwing with a cryptic error about [`waiting` being undefined](https://github.com/BryanDonovan/node-cache-manager/blob/127b83f8892f81f0903becee0b033e7ad7296d2a/lib/callback_filler.js#L13).

After debugging I found out that this was triggered by an assertion error thrown in an integration test and that error ended up in the [error handler](https://github.com/BryanDonovan/node-cache-manager/blob/127b83f8892f81f0903becee0b033e7ad7296d2a/lib/multi_caching.js#L138) of the domain created by cache manager. The failing assertion was in the callback of a method that used `.wrap` internally.

**Comment by BryanDonovan** [comment]
Thanks.  We at least need to fix the `waiting` being undefined.  I'll do that soon.  I'm not sure about if we should or how to fix the domain issue.

**Comment by diversario** [comment]
Maybe not use the domains? It's officially marked as unstable and there were/are talks about deprecating and removing them in later versions of node.

**Comment by BryanDonovan** [comment]
I like having the domains to keep the callback context, but I don't like that they catch errors.

@aletorrado - what do you think about removing the domain error handler (since you added it)?

**Comment by aletorrado** [comment]
Without domains I was having problems when the work function throws an error, because if the callback was appended into the queue the error is thrown only in the first context and the others doesn't notice.

I don't think it's possible to throw the error, or call the final callback on all items in the queue without using the domain error handler.

**Comment by danagle** [comment]
Is there a quick fix for the waiting is undefined issue? I've got a log file that is peppered with messages such as below.

{"date":"Sat Oct 17 2015 10:54:32 GMT+0000 (UTC)","process":{"pid":17557,"uid":1002,"gid":1002,"cwd":"/home/noder/app","execPath":"/usr/bin/iojs","version":"v3.3.0","argv":["/usr/bin/iojs","/usr/lib/node_modules/pm2/lib/ProcessContainerFork.js"],"memoryUsage":{"rss":120283136,"heapTotal":102065408,"heapUsed":68652504}},"os":{"loadavg":[0.00341796875,0.05859375,0.06298828125],"uptime":870707},"trace":[{"column":12,"file":"/home/noder/app/node_modules/cache-manager/lib/callback_filler.js","function":"CallbackFiller.fill","line":13,"method":"fill","native":false},{"column":36,"file":"/home/noder/app/node_modules/cache-manager/lib/caching.js","function":"","line":85,"method":null,"native":false},{"column":13,"file":"events.js","function":"emitOne","line":77,"method":null,"native":false},{"column":7,"file":"events.js","function":"Domain.emit","line":169,"method":"emit","native":false},{"column":19,"file":"domain.js","function":"Domain.errorHandler [as _errorHandler]","line":84,"method":"errorHandler [as _errorHandler]","native":false},{"column":33,"file":"node.js","function":"process._fatalException","line":208,"method":"_fatalException","native":false},{"column":24,"file":"domain.js","function":"Domain.errorHandler [as _errorHandler]","line":101,"method":"errorHandler [as _errorHandler]","native":false},{"column":33,"file":"node.js","function":"process._fatalException","line":208,"method":"_fatalException","native":false},{"column":24,"file":"domain.js","function":"Domain.errorHandler [as _errorHandler]","line":101,"method":"errorHandler [as _errorHandler]","native":false},{"column":33,"file":"node.js","function":"process._fatalException","line":208,"method":"_fatalException","native":false}],"stack":["TypeError: Cannot read property 'forEach' of undefined","    at CallbackFiller.fill (/home/noder/app/node_modules/cache-manager/lib/callback_filler.js:13:12)","    at Domain.<anonymous> (/home/noder/app/node_modules/cache-manager/lib/caching.js:85:36)","    at emitOne (events.js:77:13)","    at Domain.emit (events.js:169:7)","    at Domain.errorHandler [as _errorHandler] (domain.js:84:19)","    at process._fatalException (node.js:208:33)","    at Domain.errorHandler [as _errorHandler] (domain.js:101:24)","    at process._fatalException (node.js:208:33)","    at Domain.errorHandler [as _errorHandler] (domain.js:101:24)","    at process._fatalException (node.js:208:33)"],"level":"error","message":"uncaughtException: Cannot read property 'forEach' of undefined","timestamp":"2015-10-17T10:54:32.811Z"}

**Comment by BryanDonovan** [comment]
Hi @danagle, hopefully this branch fixes the issue for you: https://github.com/BryanDonovan/node-cache-manager/pull/36

Do you mind trying it out and letting me know if it works for you?

Thanks!

Bryan

**Comment by danagle** [comment]
Thanks - I've tested it locally and it appears to work.

Yesterday I updated the production code to check for completed jobs before executing the call and it also resolved the issue.
```javascript
redisQueue
  .getCompleted()
  .then(function(completedJobs){
    if (completedJobs.length) {
      redisQueue.clean(1000, 'completed');
    }
  });
```

**Comment by BryanDonovan** [comment]
This is fixed in 1.2.2.
